### DISCUSSION TITLE
PA 2: Correct way of using tcp_sum() and checksum values returned by it & IP address from IP header

### DISCUSSION THREAD

**Discussion by Wonigox** [ANLAB-KAIST/KENSv3#171] [Project Q&A 2. Transport Layer (TCP)]
1) I am having trouble figuring out how to use the result returned by tcp_sum().  I give the tcp_sum() function sockaddr_in.sin_adr.s_addr values for the source and destination. I also give the tcp segment header and the header length as the argument.

All these arguments are derived from the received packet in TCPAssignment::packetArrived() (assuming offset 14 for Ethernet header + 20 for IP header). However, the computed checksum returned by tcp_sum() never matches the received checksum value in the tcp header.

2) Furthermore, the IP address extracted from the IP headers (assuming that the entire Ethernet header + IP header + TCP segment headers are given) don't seem to match any packet transfers in the *.pcap files. These files only deal with the IPs 10.0.1.4 and 192.168.0.7. However, when I directly extract these addresses from the tcp segment, I get different IP addresses. 
Here is simple code on how this extraction is done:
```
  sockaddr_in src;
  sockaddr_in dst;
  
  // IP header starts from 14-th byte
  packet.readData(14+12, &src.sin_addr.s_addr, 2);
  packet.readData(14+16, &dst.sin_addr.s_addr, 2);
  src.sin_family = AF_INET;
  dst.sin_family = AF_INET;

  // TCP segment starts from 34-th byte
  packet.readData(34, &src.sin_port, 2);
  packet.readData(34+2, &dst.sin_port, 2);

  char src_ip[64] = {};
  inet_ntop(AF_INET, &src.sin_addr, src_ip, sizeof(src_ip));

  char dst_ip[64] = {};
  inet_ntop(AF_INET, &dst.sin_addr, dst_ip, sizeof(dst_ip));
  printf(----);
```

I am pretty sure there is no mistake in the extraction of IP addresses. However, the received packet sees an address that is never seen before in the *.pcap logs. As a result, the checksum is inevitably different (since, the IP addresses are different). 

I am not sure if this is caused by an error on my part since it has been stated in previous posts that our accept() is tested against the solution's connect() and our connect() is tested against the solution's accept(). This can be confirmed by the fact that our own implementation of the connect() syscall is called ~4 times while packets arrive more frequently.

Any insight would be much appreciated.

**Comment by Pusnow** [comment]
> 1. I am having trouble figuring out how to use the result returned by tcp_sum(). I give the tcp_sum() function sockaddr_in.sin_adr.s_addr values for the source and destination. I also give the tcp segment header and the header length as the argument.

The returned value should be `0xFFFF` if its checksum is correct.


> Furthermore, the IP address extracted from the IP headers (assuming that the entire Ethernet header + IP header + TCP segment headers are given) don't seem to match any packet transfers in the *.pcap files. These files only deal with the IPs 10.0.1.4 and 192.168.0.7. However, when I directly extract these addresses from the tcp segment, I get different IP addresses.
Here is simple code on how this extraction is done:

I've commented in your code

```c++
  sockaddr_in src;
  sockaddr_in dst;
  
  // IP header starts from 14-th byte
  packet.readData(14+12, &src.sin_addr.s_addr, 2); // IPv4 addresses are four bytes, so it should be 4
  packet.readData(14+16, &dst.sin_addr.s_addr, 2); // here too
  src.sin_family = AF_INET;
  dst.sin_family = AF_INET;

  // TCP segment starts from 34-th byte
  packet.readData(34, &src.sin_port, 2); // be aware that the read value is networking ordering 
  packet.readData(34+2, &dst.sin_port, 2); // here too

  char src_ip[64] = {};
  inet_ntop(AF_INET, &src.sin_addr, src_ip, sizeof(src_ip)); // after changing 2 to 4, it returns correct value.



  char dst_ip[64] = {};
  inet_ntop(AF_INET, &dst.sin_addr, dst_ip, sizeof(dst_ip));
  printf(----);
```

**Reply by Wonigox** [reply]
Oh that was a mistake on my part! 
I have a follow up question regarding checksum: I wonder why the checksum result should be 0xFFFF.

I first thought this was because the checksum field contains the received checksum as well. So I tried 0-ing out the received checksum field and computed the checksum again. 
My belief was that this should give me the same checksum value as the checksum that was sent to us. However, the two end up being different values. I don't know why computing the checksum with a 0-ed out checksum field gives a different result, even though they go through the same process?

I am not sure if this is an expected outcome or I am just doing something wrong when trying to recompute the checksum.

**Reply by Pusnow** [reply]
I'm not sure why your checksum is different. Please check your inputs (pseudo header, packet, length) are identical at byte level.
